src/parser/simplify.rs

```diff
@@ -133,8 +165,30 @@ impl ResolvableRefs for c::BodyElement {
 				// TODO
 			},
 			Rubric(e) => sub_pop(&**e, refs),
-			SubstitutionDefinition(e) => sub_pop(&**e, refs),
-			Comment(e) => sub_pop(&**e, refs),
+			SubstitutionDefinition(e) => {
+				use e::Element;
+				let subst_content = if let [c::TextOrInlineElement::ImageInline(content)] = &e.children()[..] {
+					content.as_ref()
+				} else {
+					panic!("Unexpected substitution contents.")
+				};
+				let subst = Substitution {
+					content: subst_content.clone(),
+					ltrim: e.extra().ltrim,
+					rtrim: e.extra().rtrim
+				};
+				for NameToken(name) in e.names() {
+					if refs.substitutions.contains_key(name) {
+						// TODO: Duplicate substitution name (level 3 system message).
```

Comment:
Since we first have the TargetsCollected mutable in populate_targets, then the element in resolve_refs, we need something to collect the messages into.

Reply:
Indeed. That would have to be on the element node itself, since comparing element nodes with `PartialEq` compares their contents, not their position in the tree.